machines: Make destination storage a selectable option in VM's creation

skobyda opened the pull request:
Allow user to choose already existing storage volume as VM's
destination storage. This property is optional for VM's creation.

skobyda commented:
Adds progressive disclosure option for choosing destination storage"
![Screenshot from 2019-04-08 13-36-03](https://user-images.githubusercontent.com/42733240/55721486-c7808080-5a03-11e9-990c-b65786264131.png)

When checkbox is checked, option for defining Storage Size is hidden, as this parameter becomes obsolete:
![Screenshot from 2019-04-08 13-36-22](https://user-images.githubusercontent.com/42733240/55721552-f0087a80-5a03-11e9-8128-3ccf5fcba53a.png)

Choosing destination storage is mutually exclusive with "Installation Source Type: Existing Disk Image", so when this source type is chosen, option to choose destination storage is hidden. And vice versa, if "Choose Destination Storage" checkbox is checked, "Existing Disk Image" option is hidden from the Installation Source type.
![Screenshot from 2019-04-08 13-36-35](https://user-images.githubusercontent.com/42733240/55721681-3cec5100-5a04-11e9-952b-c72becc2bd37.png)

skobyda commented:
Added "Closes: https://github.com/cockpit-project/cockpit/issues/11479" to commit message

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -314,7 +374,8 @@ class CreateVM extends React.Component {
                     { this.props.providerName == 'LibvirtDBus' &&
                     <Select.SelectEntry data={PXE_SOURCE} key={PXE_SOURCE}>{_("Network Boot (PXE)")}</Select.SelectEntry> }
-                    <Select.SelectEntry data={EXISTING_DISK_IMAGE_SOURCE} key={EXISTING_DISK_IMAGE_SOURCE}>{_("Existing Disk Image")}</Select.SelectEntry>
+                    { !this.state.chooseStorage &&
```
The logic makes sense more vise versa.
I you chose existing volumes for disk, don't show the choose storage button.. Here is someone wants to change the installation mode to choose existing disk, he has to untick the choose storage checkbox.

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -366,6 +428,53 @@ class CreateVM extends React.Component {
+
+                <hr />
+
+                { this.props.providerName == 'LibvirtDBus' &&
```
Storage Pools and volumes object exist in Virsh provider as well. You don't need to limit this feature to libvirtDBus.

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -366,6 +428,53 @@ class CreateVM extends React.Component {
+                <hr />
+
+                { this.props.providerName == 'LibvirtDBus' &&
+                this.state.sourceType != EXISTING_DISK_IMAGE_SOURCE &&
```
I believe the whole this option group should go next to the storage size. I like how virt-manager has them grouped. I would go with something similar.
![storage](https://user-images.githubusercontent.com/14921356/55787757-42ef3a00-5ab7-11e9-89af-e2b09dfe80ab.png)

skobyda replied:
After discussion with @garrett , he opted to put this option group at the bottom, as it is optional option group. What do you think @garrett ?

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -366,6 +428,53 @@ class CreateVM extends React.Component {
+                        <label className="checkbox-inline">
+                            <input id="choose-storage" type="checkbox"
+                                   checked={this.state.chooseStorage}
+                                   onChange={this.onChangedEventChecked.bind(this, 'chooseStorage')} />
```
We prefer to bind the function in the contructor to avoid binding them at each re-render.

skobyda replied:
Well, I tried to keep constistency, because rest of the component binds functions in-place.
Actually, what is the preferred way in these situations? If I add something to component, which implements stuff "and old way", should I keep consistency and then possibly change it to "new way" in another PR, or should I implement the "the new way" no matter of consistency?

KKoukiou replied:
If something 'the new way' makes performance better I would go with the new way and no care so much about consistency.

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -366,6 +428,53 @@ class CreateVM extends React.Component {
+                                               initial={this.state.storagePool}
+                                               onChange={this.onChangedValue.bind(this, 'storagePool')}>
+                                    { this.props.storagePools.map(pool => {
+                                        if (pool.connectionName === this.state.connectionName)
```
This filter should go away when you do the filtering at the outermost component for CreateVm dialog.

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -60,6 +60,25 @@ const LOCAL_INSTALL_MEDIA_SOURCE = 'file';
+    return volume ? volume.path : undefined;
+}
+
+function volumeUsedByAnotherVm(volumePath, vms) {
```
I have implemented this logic already here https://github.com/cockpit-project/cockpit/blob/master/pkg/machines/components/storagePools/storagePoolVolumesTab.jsx#L77

Please extract it out into a helper function a reuse it.

KKoukiou commented on pkg/machines/scripts/create_machine.sh:
```diff
@@ -9,6 +9,9 @@ OS="$5"
 MEMORY_SIZE="$6" # in MiB
 STORAGE_SIZE="$7" # in GiB
 START_VM="$8"
+CHOOSE_STORAGE="$9"
```
You don't need to pass the CHOOSE_STORAGE. In case STORAGE_POOL is undefined just let virt-install choose one for you.

skobyda replied:
I'm quite not sure about that. It would be quite conflicting with what "source type: existing disk image". If right user checks "choose storage", but later changes source type to existing disk image, and "choosing storage option" disappears, I have to add more handling, for example handling which would set storage_pool to undefined. It would a get a bit messy.
Right now, storage_pool keeps old value, because if choose_storage is false, then value in storage_pool is not used anyway. What do you think?

But I don't have a strong opinion on that. If you believe what I'm doing right now is wrong, I will change it.

KKoukiou replied:
You 're right. Didn't think of that. Keep it as is. Thanks.

skobyda replied:
Well, in another comment you are suggesting to always have source type option "existing disk image" at hand, and even if user chooses destination storage, if he later chooses source as existing disk image, then choose destination storage option will disappear right?
For that, I need to add handling, which will set choose_storage to false and storage_pool to undefined.

KKoukiou replied:
You can ignore the storage_pool I guess and just reset the choose_storage.

KKoukiou commented on pkg/machines/scripts/create_machine.sh:
```diff
@@ -85,13 +90,21 @@ else
 fi
 
+
+if [ "$CHOOSE_STORAGE" = "true" ]; then
```
Isn't it by default on when you pass the volume?
From man virt-install: Most checks are performed by default... Please verify that this check is not redundant.

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1258,6 +1260,20 @@ def installWithErrorTest(dialog):
+            m.execute("qemu-img create -f qcow2 /var/lib/libvirt/vmTest20.qcow2 128M")
+            m.execute("virsh pool-refresh libvirt")
+
+            createTest(TestMachines.VmDialog(self, "subVmTestCreate20", sourceType='pxe',
```
We don't use names with numbers anymore. Just leave the name field empty.

KKoukiou replied:
Please address the above comment. The commit which removed the numbering in the tests names is already merged.

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1445,6 +1465,12 @@ def fill(self):
 
+            if self.choose_storage:
+                b.wait_visible("#choose-storage")
+                b.click("#choose-storage") # TODO: fix this, do not assume initial state of the checkbox
```
You can get the checkbox value from the testing framework. The TODO comments are probably outdated.
Check wait_checked, wait_not_checked functions.

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1445,6 +1465,12 @@ def fill(self):
+            if self.choose_storage:
+                b.wait_visible("#choose-storage")
+                b.click("#choose-storage") # TODO: fix this, do not assume initial state of the checkbox
+                b.select_from_dropdown("#storage-pool-select", self.storage_pool, substring=True)
```
why you need substring=True? You have the full volume name and also the full pool name no?

KKoukiou commented:
@andreasn or @garrett could you help a bit with the grouping of the new options here?

skobyda commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -195,6 +242,19 @@ class CreateVM extends React.Component {
+            if (pool && pool.volumes) {
+                volumeEntries = (
+                    pool.volumes.map(vol => {
+                        if (vol.type === "file" || vol.type === "network" || vol.type === "block")
```
Yes after discussion with libvirt guys, they agreed other types will cause error when passed to virt-install, or are "troublesome to implement here".

KKoukiou replied:
But basically this means that virt-install does not support other pool types... Is this the case? In any case this will need to be commented with pointers to the documentation.
I mean man virt-install, search for vol will not state any limitation...

skobyda replied:
Funny thing, there is no documentation about which pool types are supported, nor any pointers.
You can only learn from libvirt code that netdir and dir volume types won't work in this case, whilst file, network and block will work. So I really don't know what comment can I leave there, apart from "these types will work".

skobyda replied:
So final decision is to do no filtering at all.

garrett commented:
Oh, when we talked about this, I meant the mutually exclusive options should switch out when the selector is changed... not to have it hidden under another checkbox.

Something like

> **Storage Pool** [ Automatic ⏷ ]

And then you select that to specify a storage pool if you do care where it goes. _Then_ the **Storage Volume** field shows up. But if you don't care where the volume goes or what it's called, it's handled by default. (I imagine there's some rule to name it without collisions, based on the OS name or somesuch.)

It's basically combining the state with the pools.

If "Automatic" doesn't work as a word, it can be changed... like "Automatic Pool & Volume", "Unspecified", etc.

KKoukiou commented:
> Oh, when we talked about this, I meant the mutually exclusive options should switch out when the selector is changed... not to have it hidden under another checkbox.
> 
> Something like
> 
> > **Storage Pool** [ Automatic ⏷ ]
> 
> And then you select that to specify a storage pool if you do care where it goes. _Then_ the **Storage Volume** field shows up. But if you don't care where the volume goes or what it's called, it's handled by default. (I imagine there's some rule to name it without collisions, based on the OS name or somesuch.)
> 
> It's basically combining the state with the pools.
> 
> If "Automatic" doesn't work as a word, it can be changed... like "Automatic Pool & Volume", "Unspecified", etc.

@garrett check also the virt-manager dialog for this.
![storage](https://user-images.githubusercontent.com/14921356/55800072-22cd7400-5ad3-11e9-86c0-ad88f439f58b.png)

garrett commented:
I don't think we should have the pool prefix in the extremely rare case someone would call their pool automatic.

I do think we should have a divider between Automatic and the pools. This should solve the "Automatic" (our label) or "Automatic" pool. If that's not enough, we could add "(pool)" after any pool called "Automatic", I guess. But, again, I really think this is an extreme corner case. I really don't understand why anyone might want to call their pool "Automatic" (or some language variant meaning the same thing).

What happens now when someone chooses "Existing disk image"? Does all the pool stuff go away like before? (If I understand it correctly, that is.)

KKoukiou commented:
> I don't think we should have the pool prefix in the extremely rare case someone would call their pool automatic.
> 
> I do think we should have a divider between Automatic and the pools. This should solve the "Automatic" (our label) or "Automatic" pool. If that's not enough, we could add "(pool)" after any pool called "Automatic", I guess. But, again, I really think this is an extreme corner case. I really don't understand why anyone might want to call their pool "Automatic" (or some language variant meaning the same thing).
> 
> What happens now when someone chooses "Existing disk image"? Does all the pool stuff go away like before? (If I understand it correctly, that is.)

@garrett I don't like how now the Storage Size and Storage Pool choice are not bound together. Do you think we can address that?

skobyda commented:
![Screenshot from 2019-04-12 12-38-42](https://user-images.githubusercontent.com/42733240/56036852-33c5f180-5d2e-11e9-8ddc-623698e915a5.png)
Bound them together with Storage Size.
Added divider and removed "pool: " prefix.

KKoukiou commented on pkg/machines/scripts/create_machine.sh:
```diff
@@ -85,13 +89,21 @@ else
 
+
+if [ "$STORAGE_POOL" != "Automatic" ]; then
+    CHECK_PARAM="--check path_in_use=off"
```
Why you need to pass this? I can imagine that you expect the volume you pass to always exist so you don''t need virt-install to do the check for you. But if this doesn't hurt to not be specified I 'd rather remove it.

skobyda replied:
That's not what this parameter does. What it does it it checks if disk is already used by another VM. Default value (on) throws error if disk is already used. (see man virt-install)
However this PR allows user to also choose storage, which is used by another VM. So we want the parameter "--check path_in_use=off" in case "custom storage" was specified by user.

KKoukiou replied:
Oh right, I got mislead.

Generally, I would suggest introducing the changes one by one.

The support for storage pool/volumes choice would make sense by itself, without allowing multiple VMs to use the same pool. You don't need to squash everything in a single commit. 

In addiction this --check path_in_use=off newly introduced option affects not only your newly introduced installation option choice, but all existing ones as well.

For example, if someone chooses 'Use existing disk option' now, he will not see any warning that some other VM is using that disk, (because this was never implemented) but also virt-install will not fail, because you removed the check here. This can lead in a user starting a new VM with a used disk, resulting in corruption in the disk, which will be mounted rw in two VMs.

skobyda replied:
Is mentioned in IRC, --check path_in_use is set to off only for the functionality of this PR (that's what [ "$STORAGE_POOL" != "Automatic" ] does). Otherwise it will say on default value, therefore not affecting anything outside of this PR I believe.

KKoukiou commented on pkg/machines/scripts/install_machine.sh:
```diff
@@ -73,6 +73,7 @@ virt-install \
     --wait -1 \
     --noautoconsole \
     --noreboot \
+    --check path_in_use=off \
```
This needs explanation. It's not only affecting what this PR is supposed to do.

skobyda replied:
Actually, this parameter is useful only when supplying already existing disk (what this PR does).
It won't affect anything but this PR, and It won't change behavior in any other use cases, since in all the other use cases we create our own disks for every VM creation, so checking if disk is already used by another VM is useless.
But I can leave comment "allow user tor supply disk used by another VM" if you want.

KKoukiou replied:
What about when importing disks...? That option used to exist previously as well.

skobyda replied:
for importing disk, that will be prevented in create_machine.sh, so I guess there is no need to also check it in install_machine.sh ?

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1258,6 +1258,19 @@ def installWithErrorTest(dialog):
+                                             os_name=config.NOVELL_NETWARE_6,
+                                             storage_pool="libvirt",
+                                             storage_volume="vmTest20.qcow2",
+                                             start_vm=False,))
```
I would rather have here start_vm=True. It performs a completely logic test. We already test starting VMs from local images in subVmTestCreate17.
Also if you can add a test for separate install step that would be awesome!

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1445,6 +1461,12 @@ def fill(self):
                 b.click("#start-vm") # TODO: fix this, do not assume initial state of the checkbox
             # b.set_checked("#start-vm", self.start_vm)
 
+            if self.storage_pool != 'Automatic':
```
Can you add a line to check that if Automatic is set for pool the volume entry is absent?

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -350,14 +394,44 @@ class CreateVM extends React.Component {
                                  onValueChange={this.onChangedEventValue.bind(this, 'storageSize')}
                                  onUnitChange={this.onChangedValue.bind(this, 'storageSizeUnit')} />}
 
+                <label className="control-label" htmlFor="storage-pool">
```
the htmlFor id is different than the id of the select

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -350,14 +394,44 @@ class CreateVM extends React.Component {
+
+                { this.state.storagePool !== "Automatic" &&
+                <React.Fragment>
+                    <label className="control-label" htmlFor="storage-volume">
```
Mismatch htmlFor as above.

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -350,14 +394,44 @@ class CreateVM extends React.Component {
                                  onValueChange={this.onChangedEventValue.bind(this, 'storageSize')}
                                  onUnitChange={this.onChangedValue.bind(this, 'storageSizeUnit')} />}
 
+                <label className="control-label" htmlFor="storage-pool">
```
The whole storage pool, volume, size section should not be visible when the user selects to use existing disk image.

KKoukiou replied:
Also I tend to believe that the storage size, storage pool, storage volume fields should get separated from the memory field, but that's up to @garrett  to decide.

garrett commented:
It's really hard for me to look at the create new VM dialog without wanting to redo everything, especially since it's all interconnected.

- Installation isn't always installation — such as PXE boot (which might or might not be installation) or existing disk image (which is already installed)
- OS Vendor and Operating System are almost meaningless
  - what really matters is if it's a modern OS (with virtio, etc.), Windows, or legacy support (IDE, etc.)… 
  - aside from that, I guess there might be some corner cases like BSDs, Haiku, etc. — but these probably fall under one of the 3 categories (depending on where, we might want to think about catigorizations of Linux, Non-Linux, Legacy? or some other scheme)
- Storage should all fit together as one group of widgets
- Name should be the first item, not connection...
  - Connection is actually where this the VM is stored... system-wide or just for the user
  - Move "connection" into destination

So, perhaps the groups of items in the dialog would look like this:

- Name
- Source
- Destination
- Settings
- Auto-start

And it would be sort of like this with more details:

  |  
--: | --
**Name** | _____
  |  
**Source** | [DROPDOWN]
**Source_Label_Here** | (Convert placeholders to short labels for this field)
  |  
**Destination** | System / User
**Pool** | [Automatic / Storage Pools here]
**Volume** | [Dropdown] / Generated volume name, in an `<output>`
**Size** | (Only shows when applies)
  |  
**Memory** | Memory size here
**Compatibility** | [Modern / Windows / Legacy IDE]
  |  
  | Auto-start

And then memory and compatibility would have good generic defaults for modern Linux distros and automatically changed to an alternate if it is detected that a non-Linux OS is being used.

garrett commented:
Note: I took a step back to think about how it would all fit together better. I don't intend for all of my suggestions in the previous comment to go into this PR — it's only to provide a goal on where we might go with the dialog.

What I would like to see in this PR is the related parts rough design above — that is, the destination section, basically. The rest can go into another PR... which would hopefully land shortly after this one, in the same Cockpit release version.

garrett commented:
How's this?

I dialed back the changes a bit with respect to the OS vendor/version stuff.

One larger change was to redesign the VM type switcher to be a group of radio buttons for system or user, instead of a dropdown with QEMU/KVM jargon. 

I didn't add what happens when user is selected (like we talked about yesterday)... but we don't do that now either. (Ideally, we'd communicate what restrictions there would be when someone creates a user VM instead of a system VM.... _or_ we'd get rid of being able to create user VMs entirely.)

I'm fine with keeping the system/user switcher as-is for this PR, as long as we do address it afterward.

![create-vm-storage](https://user-images.githubusercontent.com/10246/56199184-b0addf80-603c-11e9-9622-0edf91710ce7.png)

skobyda commented:
@garrett Thanks for mockup, it looks great.
For this PR, I redesigned only stuff related to Storage Pools (and pulled Memory into separate Group).
Other changes are outside of this PR, and I will create a new PR right after this one gets merged, where I will refactor and redesigned all the dialog according to your suggestion

![Screenshot from 2019-04-16 13-18-30](https://user-images.githubusercontent.com/42733240/56205561-721f2180-604a-11e9-8c87-c6e73cd6f15b.png)
![Screenshot from 2019-04-16 13-18-57](https://user-images.githubusercontent.com/42733240/56205564-73e8e500-604a-11e9-8c76-177571be0385.png)
.

garrett commented:
> Other changes are outside of this PR, and I will create a new PR right after this one gets merged

Cool, thanks! :+1:

Changing the label for "Storage Pool" And "Storage Volume" makes sense to do so too, but that can wait for the next PR also, if you think so... provided this and the next PR are both in the same release. :wink:

skobyda commented:
@garrett
"Storage Pool" -> "Storage"
"Storage Volume" -> "Volume"
"Storage Size" -> "Size"

skobyda commented:
Ups, didn't notice conflict with master, rebased

KKoukiou commented on pkg/machines/scripts/create_machine.sh:
```diff
@@ -85,13 +89,21 @@ else
 fi
 
+
+if [ "$STORAGE_POOL" != "CreateVolume" ]; then
```
createVolume? I can't seem to find this anywhere. Did you mean NewVolume?

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1264,6 +1264,20 @@ def installWithErrorTest(dialog):
+            self.machine.execute("virsh pool-refresh libvirt")
+
+            # Check choosing existing volume as destination storage
+            createTest(TestMachines.VmDialog(self, "subVmTestCreate20", sourceType='pxe',
```
Please don't use PXE. The test is very slow, just choose another installation method.

garrett commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -380,15 +423,40 @@ class CreateVM extends React.Component {
+                <Select.Select id="storage-pool-select"
+                               initial={this.state.storagePool}
+                               onChange={e => this.onChangedValue('storagePool', e)}
+                               enabled={this.state.sourceType != EXISTING_DISK_IMAGE_SOURCE}>
```
This is part of the reason why I shuffled everything around in the mockups — so we can have source & destination next to each other so they have a bit more context when things go away.

While some of the design suggestions should probably still be in another PR (such as adjusting label text, changing the system connection from a dropdown to radio buttons), changing the order probably should be in this PR so storage makes sense when this PR lands.

skobyda replied:
So I moved destination next to the source.
Also now i somebody chooses source: existing disk image, then all the destination option disappear instead of just being disabled.
![Screenshot from 2019-04-24 10-57-03](https://user-images.githubusercontent.com/42733240/56646439-db70e700-667f-11e9-8ace-591df085e062.png)
![Screenshot from 2019-04-24 10-57-12](https://user-images.githubusercontent.com/42733240/56646446-ddd34100-667f-11e9-9433-402b3206a0fe.png)

garrett replied:
Much better! Thanks!

Looking forward to the other items getting resorted too. (In the follow-up PR.)

skobyda commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -155,6 +158,13 @@ class CreateVM extends React.Component {
+
+            // this source type is incompatible with "choose destination storage" option
+            if (value === EXISTING_DISK_IMAGE_SOURCE) {
+                this.setState({ storagePool: "NewVolume" });
```
If EXISTING_DISK_IMAGE_SOURCE, then NewVolume won't be used in bash script. This is related to the discussion below (about Destiantion dropdown being disabled, but not greyed out). So if you choose existing disk image source, that destination is automatically set to new volume and disabled. However now I see it makes more sense to just "disappear".

KKoukiou replied:
I believe my comment is valid and this is redundant here. If you choose EXISTIN_DISK_IMAGE the whole custom storage section will disappear now, so having preset values for these is not required.

skobyda replied:
Yeah right, that setting to default value should be removed.

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -358,6 +407,49 @@ class CreateVM extends React.Component {
+                    </React.Fragment> }
+
+                    { this.state.storagePool === "NewVolume" &&
+                    <MemorySelectRow label={_("Size")}
```
label is ignored in MemorySelectRow component now, you have to put it outside, otherwise label is missing.

KKoukiou replied:
Needs:
```
<label htmlFor='storage-size' className='control-label'>         
  {_("Size")}                                               
  </label>                                                        
```

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -380,15 +423,40 @@ class CreateVM extends React.Component {
+                    <Select.SelectEntry data="NewVolume" key="NewVolume">{"Create New Volume"}</Select.SelectEntry>
+                    <Select.SelectDivider />
+                    <optgroup key="Storage Pools" label="Storage Pools">
+                        { this.props.storagePools.map(pool => <Select.SelectEntry data={pool.name} key={pool.name}>{pool.name}</Select.SelectEntry>)}
```
This is not fixed but it's resolved.

skobyda replied:
I don't understand what do you mean. If someone changes the connection type, then state.connectionName of CreateVmModal component changes. Then React will rerender (I dunno if that's the proper word) everything that uses state.connectionName. storagePools list (line 624), which gets passed to CreateVM component, uses it, so list will be rerendered.

KKoukiou replied:
I was confused because on my system setup I had exactly the same names for system and session pools. Sorry :/

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -358,6 +401,54 @@ class CreateVM extends React.Component {
+                        <label htmlFor='storage-size' className='control-label'>
+                            {_("Size")}
+                        </label>
+                        <MemorySelectRow label={_("Size")}
```
The label is ignored so please remove it.

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1162,8 +1162,32 @@ def installWithErrorTest(dialog):
+            createTest(TestMachines.VmDialog(self, sourceType='file',
+                                             location=config.NOVELL_MOCKUP_ISO_PATH,
+                                             memory_size=50, memory_size_unit='MiB',
+                                             storage_size=0, storage_size_unit='MiB',
```
I think this is redundant since no storage is selcted.

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1162,8 +1162,32 @@ def installWithErrorTest(dialog):
+            createTest(TestMachines.VmDialog(self, sourceType='file',
+                                             location=config.NOVELL_MOCKUP_ISO_PATH,
+                                             memory_size=50, memory_size_unit='MiB',
+                                             storage_size=0, storage_size_unit='MiB',
```
We don't choose size when choosing specific volume this this should be removed as well.

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1162,8 +1162,32 @@ def installWithErrorTest(dialog):
+            self.machine.execute("qemu-img create -f qcow2 /var/lib/libvirt/vmTestCustomDestination.qcow2 128M")
+            self.machine.execute("virsh pool-refresh libvirt")
+
+            # Check choosing existing volume as destination storage
```
The following two tests are exposed with the virsh provider for users, thus testing them should not be limited to libvirt-dbus being installed.

KKoukiou commented on test/verify/machineslib.py:
```diff
@@ -1188,13 +1188,21 @@ def installWithErrorTest(dialog):
         createTest(TestMachines.VmDialog(self, sourceType='file',
                                          location=config.NOVELL_MOCKUP_ISO_PATH,
                                          memory_size=50, memory_size_unit='MiB',
-                                         storage_size=0, storage_size_unit='MiB',
```
This got introduced in the previous commit, it should not be inserted there anyway.

KKoukiou commented:
@skobyda this still has errrors in the testCreate, thus keeping needwork label. Please fix.
```
warning: --: 29: [: tmpPool: unexpected operator

warning: GET_VM action failed failed for path /org/libvirt/QEMU/domain/_99327d7f_5ca0_420c_8164_aeac9afd4558 
warning: --: 29: [: NoStorage: unexpected operator
ERROR    Error: --disk vol=NoStorage/undefined: Storage pool not found: no storage pool with matching name 'NoStorage'
```

KKoukiou commented:
For release notes:
![diff-pool-create-vm](https://user-images.githubusercontent.com/14921356/57891064-591cb280-783a-11e9-861e-0c3e9ab2d09f.png)

KKoukiou commented on pkg/machines/helpers.js:
```diff
@@ -602,18 +602,17 @@ function getVmDisksMap(vms, connectionName) {
- * Finds, if volumes in a storage pool are used by any VM. If so
- * name of VM is appended to dictionary, which contains arrays of all VMs
- * using a certain pool.
+ * Returns a object of key-value pairs of Storage Volume names mapping
```
I fixed it with the last push.

KKoukiou commented on pkg/machines/components/create-vm-dialog/createVmDialog.jsx:
```diff
@@ -518,20 +602,26 @@ class CreateVmModal extends React.Component {
 
                 <hr />
 
+                { this.state.sourceType != EXISTING_DISK_IMAGE_SOURCE &&
```
Fixed with last push.

dacresni commented:
but how do you create a new volume in an existing stoage pool?

skobyda commented:
> but how do you create a new volume in an existing stoage pool?

Just got PR merged which implements creation of volumes: https://github.com/cockpit-project/cockpit/pull/12589
